STORM-1136: Command line module to return kafka spout offsets lag and…

priyank5485 opened the pull request:
… display in storm ui.

HeartSaVioR commented:
@priyank5485 Could you attach UI screenshot to see output easily? Thanks in advance.

priyank5485 commented:
![kafkaspoutlags_success](https://cloud.githubusercontent.com/assets/5192436/15682583/402a1180-2713-11e6-9ad2-e101cbb1ea4c.png)
![kafkaspoutlags_failure](https://cloud.githubusercontent.com/assets/5192436/15682586/43807626-2713-11e6-8609-870774dbc511.png)

priyank5485 commented:
@HeartSaVioR I just attached the two screen shots. One when it successfully managed to get the lag result. Other one by stopping kafka in which case it will just show a message in the UI indicating the cause.

harshach commented:
@priyank5485 Nice!!. Can we rename the section as KafkaSpout Lag

priyank5485 commented:
@harshach Will do that. Just to confirm, you mean the section header in UI under Topology Stats, right? Also, since the topology can technically have more than one kafka spout and we are showing lags for all spouts there I was thinking of keeping it as plural. May be Kafka Spouts Lag

Does that sound okay?

harshach commented:
@priyank5485 Yes. sounds good.

priyank5485 commented:
@harshach Done.

harshach commented on external/storm-kafka-monitor/src/main/java/org/apache/storm/kafka/monitor/KafkaOffsetLagResult.java:
```diff
@@ -0,0 +1,102 @@
+/*
```
can we put this in org.apache.storm.tools.KafkaOffsetLagTool.java

harshach commented on external/storm-kafka-monitor/src/main/java/org/apache/storm/kafka/monitor/KafkaOffsetQueryInfoNew.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Class representing information for querying kafka for log head offsets, consumer offsets and the difference for new kafka spout using new consumer api
+ */
+public class KafkaOffsetQueryInfoNew {
```
Should we call this as NewKafkaSpoutOffsetQuery and OldKafkaSpoutOffsetQery

harshach commented on external/storm-kafka/src/jvm/org/apache/storm/kafka/KafkaSpout.java:
```diff
@@ -188,6 +189,47 @@ public void declareOutputFields(OutputFieldsDeclarer declarer) {
     }
 
+    @Override
+    public Map<String, Object> getComponentConfiguration () {
```
looks to be quite a bit of logic in here. If it makes sense lets move most of it into KafkaUtils.java

harshach commented:
@priyank5485 this looks great. Left some minor nit-picks.

darionyaphet commented:
LGTM

priyank5485 commented:
@harshach I have renamed Kafka Log Head Offset to Latest Offset, since Kafka is present in ui header and topic is a column already. I have also renamed the Query classes as per your suggestion.

Regarding moving it under package tool, I thought of this module specific to kafka and getting the offset lags. If you are okay with it we can keep it as is instead of creating a tools package. Or any other suggestion if you have.

Regarding moving getComponentConfiguration to KafkaUtils, KafkaUtils exist only in the old module. So i will keep it in the spout for now. I plan to add getComponentConfiguration for trident as well. If there is code duplication there, I will try to reuse it for old module and new module(when trident is ready).

Let me know if there is anything else you want me to address. Otherwise I will squash the commits

harshach commented:
@priyank5485 I overlooked that this code is in external/module we can keep the classes as it is. My comment regarding getComponentConfiguration as it looks to be doing bit of logic in KafkaSpout and my suggestion was to move that code to KafkaUtils and call buildKafkaSpoutComponentConfig and return the config map . This way spout code looks cleaner and may be buildKafkaSpoutComponentConfig can be reused for trident. 
Go ahead with you commit squashes and you can add me to sponsor list since its in external module

hmcl commented on external/storm-kafka-monitor/src/main/java/org/apache/storm/kafka/monitor/KafkaOffsetLagUtil.java:
```diff
@@ -0,0 +1,374 @@
+ * Utility class for querying offset lag for kafka spout
+ */
+public class KafkaOffsetLagUtil {
+    private static final String OPTION_TOPIC_SHORT = "t";
```
what is the difference between SHORT and LONG ? why t vs topics ?

priyank5485 replied:
I think its just convenience. You can use -t or --topics.

hmcl commented on external/storm-kafka-monitor/src/main/java/org/apache/storm/kafka/monitor/KafkaOffsetLagUtil.java:
```diff
@@ -0,0 +1,374 @@
+                }
+                result.get(leaders[i]).add(new TopicPartition(oldKafkaSpoutOffsetQuery.getTopic(), Integer.parseInt(partitions[i])));
+            }
+        } else {
```
when can this scenario occur? There is a lot of code in some of these methods. It would be ideal to put some comments hinting on what is going on, and/or split into smaller methods.

priyank5485 replied:
Old kafka spout has two ways for its configuration. 
1. StaticHosts
2. ZkHosts
StaticHosts is when user adds the leader for each partition statically at the time of building the topology. In that case we need to use that information to get latest offsets. For ZkHosts the KafkaSpout finds leaders in the prepare method using zk nodes and consumes data. Thats what the if else represents. I will add some documentation.

harshach commented:
Thanks @priyank5485 . This looks great and will be greatly helpful for all the users of KafkaSpout. Lets open another JIRA for trident support. +1.

priyank5485 commented:
opened https://issues.apache.org/jira/browse/STORM-1891to track trident kafka spouts

abellina commented on storm-core/src/ui/public/topology.html:
```diff
@@ -384,8 +389,51 @@ <h2 id="topology-resources-header">Topology resources</h2>
+                  return true;
+                };
+                var kafkaSpoutsValidResults = kafkaSpoutLags.filter(function (ele) {return isJson(ele.spoutLagResult);});
+                var kafkaSpoutsErrorResults = kafkaSpoutLags.filter(function (ele) {return !isJson(ele.spoutLagResult);});
```
Could the response still be json in the error case? I think the response should be JSON at all times.

priyank5485 replied:
@abellina If its an error we only want to show an appropriate message to the user as a string. Also, not that the top level response is still a json. I think string is fine here.

abellina commented on storm-core/src/ui/public/topology.html:
```diff
@@ -384,8 +389,51 @@ <h2 id="topology-resources-header">Topology resources</h2>
+                  data.kafkaSpoutsLagResults = [];
+                  kafkaSpoutsValidResults.forEach(function(ele) {
+                    var spoutLagResult = JSON.parse(ele.spoutLagResult);
+                    spoutLagResult.forEach(function(ele2) {
```
Could the rest endpoint return id instead of spoutId, and this loop can go away?

abellina replied:
or it returns spoutId (as it does now), and the rest of the code is adapted to take spoutId.

priyank5485 replied:
We would still need the for loop as need to flatten the results out. I chose to use id in the json passed to template since the template uses id in other tables for ui. If you are fine we can leave it as is.

abellina replied:
@priyank5485 The result currently looks like (if I understand correctly):

[{spoutId: "spout-1", spoutLagResult: "{topic: '..', partition: '..', logHeadOffset: '..', ...}",
 {spoutId: "spout-2", spoutLagResult: "error message"]

I think this should be split into two arrays. Sample proposed response:

// notice no strings, it is strictly JSON.
 {valid: [{spoutId: "spout-1", spoutLagResult: [topic: '..', partition: '..', logHeadOffset: '..', ...]}, ... ],
  invalid: [{spoutId: "spout-2", error: ""}, ...]}

This tells you at a glance if something had an error: resp.invalid.length > 0, tells you if there are non-errors spouts: resp.valid.length > 0, and allows you to loop through resp.valid and resp.invalid to populate your tables.

priyank5485 replied:
@abellina The spoutLagResult property is actually an array. Hence the flattening. 

Let me elaborate the limitations of valid and invalid approach. The TopologySpoutLag class in storm-core called by ui server does not know anything about if the spoutLagResult is valid or not. Reason is,  the way to get lag information for kafka and other spouts is handled by making a shell call handled by an external module since we did not want any direct dependency on storm core. For now we have kafka, but we can add other types of spouts as well. Plus the fields in ui or the template for a different type of spout could be different. The only commonality I found was if something went wrong for getting lag info for underlying spout(kafka or other) then an error message is sent in response. This will work for all types of spouts. To do what you are saying we will have to inspect the response in TopologySpoutLag class. I felt it did not matter if it were ui doing that or server. I preferred ui. Let me know if it makes sense or not and if you still think we need to change something.

abellina replied:
@priyank5485 yes, sorry that was my intention but I didn't write it correctly in my comment, spoutLagResult should be a JSON array. Thanks for the background!

That said, I think that TopologySpoutLag should do this (my two cents). That way the output from the rest interface doesn't need further inspection and some other script can call on your new lag endpoint and get a JSON object they use without having to do the extra testing. I am not sure that valid/invalid are the best labels here by the way, just an example I came up with.

abellina commented:
Could the new tables be made optional? That way when a topology doesn't use KafkaSpout, the tables don't show in the UI.

priyank5485 commented:
@abellina The tables are optional already. If there are no kafka spouts then the response will be an empty array. Can you try running a non kafka spout topology and check? I did.

priyank5485 commented:
@harshach Rebased 1.x branch since there were some merges since the PR was raised.

abellina commented on storm-core/src/ui/public/topology.html:
```diff
@@ -384,8 +389,51 @@ <h2 id="topology-resources-header">Topology resources</h2>
+                var kafkaSpoutsLagTemplate = $(template).filter("#topology-kafka-spouts-lag-template").html();
+                var spoutsErrorTemplate = $(template).filter("#topology-spouts-lag-error-template").html();
+
+                var kafkaSpoutLags = lagResponse.filter(function(ele) {return ele.spoutType === "KAFKA";});
```
is the /lag endpoint returning spouts that aren't KafkaSpout?

harshach replied:
@abellina Yes thats the idea to extend it for other spouts. For now only kafkaspout supported

abellina commented:
@priyank5485 Yes I see the tables don't get populated if lag responses are empty, but we still end up with the div markup on the page. It seems that this may be an area where other spouts could put their stats in the future. Could you call the div topology-spout-stats, and put a comment in the html that says it gets populated if stats are available?

priyank5485 commented:
@abellina I have added a comment for that div. As far as renaming it to topology-spout-stats is concerned I prefer the current name since right below that div there is a div showing stats for the spout like complete latency, etc. I dont want the two to be confused.

harshach commented:
sitll +1. @abellina can you address @priyank5485 response.

abellina commented:
@harshach I added a comment on the format for the response from the lag endpoint. I still believe the backend should do some of the work the UI is currently doing. 

@priyank5485 div name makes sense now. Thanks!

harshach commented:
@abellina lets handle the concern you've in https://issues.apache.org/jira/browse/STORM-1891.
